### PR TITLE
SOLR-11749 Several test changes

### PR DESCRIPTION
**Pull request by madrob**
https://issues.apache.org/jira/browse/SOLR-11749

Still some rough edges, it's all macOS only for now, but I think it's a good start.

### CODE REVIEW

File: solr/docker/tests/cases/initdb/test.sh [hunk 1]
```diff
@@ -41,7 +41,7 @@ if [[ "$data" != /var/solr/initdb-was-here ]]; then
   echo "Test $TEST_DIR $tag failed; script did not run"
   exit 1
 fi
-data=$(docker exec --user=solr "$container_name" ls /var/solr/should-not-be; true)
+data=$(docker exec --user=solr "$container_name" ls /var/solr/should-not-be ; true)
```
Comment (HoustonPutman) [inline solr/docker/tests/cases/initdb/test.sh:44]: ?

File: solr/packaging/build.gradle [hunk 2]
```diff
@@ -160,3 +165,55 @@ distTar {
...
+  outputs.dir("${project.ext.nodeProjectDir}/node_modules/bats")
+  outputs.dir("${project.ext.nodeProjectDir}/node_modules/bats-support")
+  outputs.dir("${project.ext.nodeProjectDir}/node_modules/bats-assert")
+}
+
+task bats(type: Exec) {
+  dependsOn installDist
+  dependsOn downloadBats
+
+  def testOutput = "$buildDir/bats-output"
```
Comment (HoustonPutman) [inline solr/packaging/build.gradle:195]: Can we call this `integration-test-output`, and rename the task to `integrationTests` or something like that. To me, most people won't know what Bats is.

File: solr/packaging/test/bats_helper.bash [hunk 3]
```diff
@@ -0,0 +1,65 @@
+#!/usr/bin/env bats
```
Comment (HoustonPutman) [inline solr/packaging/test/bats_helper.bash:1]: is this a bash or bats file?

**Review by HoustonPutman — CHANGES_REQUESTED** [review]

File: solr/packaging/test/README.md [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+ characters like dashes or underscores, so JIRA issue numbers
+ and command line flags are not valid test names. For more detail
+ about BATS features, please consult 
+ special features of BATS, please consult 
+
+Some tests will start clusters or create collections,
+ please take care to delete any resources that you create.
+ They will not be cleaned for you automatically.
+
+It is recommended that you install and run `shellcheck` to verify your test scripts and catch common mistakes before committing your changes.
```
Comment (gerlowskija) [inline solr/packaging/test/README.md:49]: Neat!

File: solr/packaging/test/test_start_solr.bats [hunk 5]
```diff
@@ -14,12 +15,19 @@
...
+
+teardown() {
+  solr stop -all >/dev/null 2>&1
+}
+
+@test "SOLR11740 check f" {
+  run -0 solr start
+  run -0 solr start -p 7574
+  run bash -c 'solr stop -all 2>&1'
+  refute_output --partial 'forecfully killing'
```
Comment (gerlowskija) [inline solr/packaging/test/test_start_solr.bats:32]: [Q] Is 'forecfully' a typo here, or a part of the test?

File: solr/packaging/test/README.md [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+
+# bin/solr Tests
+
+This directory contains tests for the `bin/solr` command-line scripts.
+
+All tests in this project use the [BATS](https://bats-core.readthedocs.io/en/stable/index.html) framework.
+
+## Running Tests
+
+Tests can be run via `./gradlew bats`.
```
Comment (gerlowskija) [inline solr/packaging/test/README.md:26]: [-1] Should the task name here be `integrationTests`?  That's the task that works for me from the CLI, and the one that I see in `solr/packaging/build.gradle`

**Review by gerlowskija — COMMENTED** [review]
LGTM - bats is a big improvement over the handrolled setup this was using before.

Tests look good on my Mac.  I'm going to try Linux now and will update this comment to an approval assuming that passes.

**Comment by gerlowskija** [comment]
LGTM - it passed on my Macbook and on an Ubuntu 20.04 machine I have.

**Review by HoustonPutman — APPROVED** [review]
Looks great, awesome work 🙂

**Comment by HoustonPutman** [comment]
This does likely warrant a changes.txt entry. (And you might want to add a github action to run the tests when only that directory has been changed, we can expand it later)

**Comment by madrob** [comment]
re: CHANGES.txt

@janhoy - 9.0 or 9.1?

**Comment by janhoy** [comment]
Nice work. I’m ok with 9.0, but be prepared to fix failing tests, or to revert if it becomes unstable for some other reason.

**Comment by madrob** [comment]
> be prepared to fix failing tests

Part of the problem is that we weren't running any of these tests before!

File: solr/bin/solr [hunk 7]
```diff
@@ -99,14 +102,13 @@ elif [ -r "$SOLR_INCLUDE" ]; then
...
-fi
+# if pid dir is unset, deafult to $solr_tip/bin
+: ${SOLR_PID_DIR:=$SOLR_TIP/bin}
 
-if [ -n "$SOLR_JAVA_HOME" ]; then
+if [ -n "${SOLR_JAVA_HOME:-}" ]; then
   JAVA="$SOLR_JAVA_HOME/bin/java"
   JSTACK="$SOLR_JAVA_HOME/bin/jstack"
-elif [ -n "$JAVA_HOME" ]; then
+elif [ -v JAVA_HOME ]; then
```
Comment (cpoerschke) [inline solr/bin/solr:111]: This gives `[: -v: unary operator expected` for older versions of bash.

Could `-n "${JAVA_HOME:-}"` be used instead, consistent with the `SOLR_JAVA_HOME` check a few lines above?

Comment (madrob) [reply solr/bin/solr:111]: Done in c3b56ab, thank you.

**Pull request merged**